hydrolib/core/io/bc/models.py

```diff
@@ -220,23 +271,177 @@ class T3D(ForcingBase):
+
+    _key_renaming_root_validator = get_key_renaming_root_validator(
+        {
+            "timeinterpolation": ["time_interpolation"],
+            "vertpositions": ["vertical_position_specification"],
+            "vertinterpolation": ["vertical_interpolation"],
+            "vertpositiontype": ["vertical_position_type"],
+            "vertpositionindex": ["vertical_position"],
+        }
+    )
 
     _split_to_list = get_split_string_on_delimiter_validator(
-        "verticalpositions",
+        "vertpositions",
     )
 
     _verticalinterpolation_validator = get_enum_validator(
-        "verticalinterpolation", enum=VerticalInterpolation
+        "vertinterpolation", enum=VerticalInterpolation
     )
     _verticalpositiontype_validator = get_enum_validator(
-        "verticalpositiontype", enum=VerticalPositionType
+        "vertpositiontype", enum=VerticalPositionType
+    )
+    _timeinterpolation_validator = get_enum_validator(
+        "timeinterpolation", enum=TimeInterpolation
     )
 
+    @root_validator(pre=True)
+    def _validate_quantityunitpairs(cls, values: Dict) -> Dict:
+        super()._validate_quantityunitpair(values)
+
+        quantityunitpairs = values["quantityunitpair"]
+
+        T3D._validate_that_first_unit_is_time_and_has_no_verticalposition(
+            quantityunitpairs
+        )
+
+        verticalpositions = values.get("vertpositions")
+        if verticalpositions is None:
+            raise ValueError("vertpositions is not provided")
+        verticalpositionindexes = values.get("vertpositionindex")
+
+        number_of_verticalpositions = (
+            len(verticalpositions)
+            if isinstance(verticalpositions, List)
+            else len(verticalpositions.split())
+        )
+
+        if verticalpositionindexes is None:
+            T3D._validate_that_all_non_time_quantityunitpairs_have_valid_verticalpositionindex(
+                quantityunitpairs, number_of_verticalpositions
+            )
+            return values
+
+        T3D._validate_verticalpositionindexes_and_update_quantityunitpairs(
+            verticalpositionindexes,
+            number_of_verticalpositions,
+            quantityunitpairs,
+        )
+
+        return values
+
+    @staticmethod
+    def _validate_that_first_unit_is_time_and_has_no_verticalposition(
+        quantityunitpairs: List[QuantityUnitPair],
+    ) -> None:
+        if quantityunitpairs[0].quantity.lower() != "time":
+            raise ValueError("First quantity should be `time`")
+        if quantityunitpairs[0].vertpositionindex is not None:
+            raise ValueError("`time` quantity cannot have vertical position index")
+
+    @staticmethod
+    def _validate_that_all_non_time_quantityunitpairs_have_valid_verticalpositionindex(
+        quantityunitpairs: List[QuantityUnitPair], maximum_verticalpositionindex: int
+    ) -> None:
+        for quantityunitpair in quantityunitpairs:
```

Comment:
```suggestion
        for quantityunitpair in quantityunitpairs[1:]:
```

and then you can remove

```
            if quantity == "time":
                continue
```

Reply:
Yes, I like it. Done!